File: backends/node/shared/implementations/middleware/idempotency/default/src/handlers/check.ts
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { DistributedCache, Idempotency } from "@d2/interfaces";
import { z } from "zod";
import { DEFAULT_IDEMPOTENCY_OPTIONS, type IdempotencyOptions } from "../idempotency-options.js";

type CheckInput = Idempotency.CheckInput;
type CheckOutput = Idempotency.CheckOutput;

const SENTINEL = "__processing__";
const KEY_PREFIX = "idempotency:";

/**
 * Handler for checking idempotency state using SET NX + GET pattern.
 * Mirrors D2.Shared.Idempotency.Default.Handlers.X.Check in .NET.
 *
 * Complex handler (X/) because it may mutate shared state (creates Redis keys)
 * while its primary intent is retrieval of state.
 *
 * Fail-open on all cache errors.
 */
export class Check
  extends BaseHandler<CheckInput, CheckOutput>
  implements Idempotency.ICheckHandler
{
  private static readonly checkSchema = z.object({
    idempotencyKey: z.string().min(1).max(256),
  }) as unknown as z.ZodType<CheckInput>;

  private readonly setNx: DistributedCache.ISetNxHandler<string>;
  private readonly get: DistributedCache.IGetHandler<string>;
  private readonly options: IdempotencyOptions;

  constructor(
    setNx: DistributedCache.ISetNxHandler<string>,
    get: DistributedCache.IGetHandler<string>,
    options: Partial<IdempotencyOptions>,
    context: IHandlerContext,
  ) {
    super(context);
    this.setNx = setNx;
    this.get = get;
    this.options = { ...DEFAULT_IDEMPOTENCY_OPTIONS, ...options };
  }

  protected async executeAsync(input: CheckInput): Promise<D2Result<CheckOutput | undefined>> {
    // Validate input.
    const validation = this.validateInput(Check.checkSchema, input);
    if (validation.failed) {
      return D2Result.bubbleFail(validation);
    }

    const cacheKey = `${KEY_PREFIX}${input.idempotencyKey}`;

    try {
      // 1. Attempt to acquire the lock with SET NX.
      const setNxResult = await this.setNx.handleAsync({
        key: cacheKey,
        value: SENTINEL,
        expirationMs: this.options.inFlightTtlMs,
      });

      if (setNxResult.success && setNxResult.data?.wasSet === true) {
        // Lock acquired — caller should proceed with request processing.
        return D2Result.ok({
          data: { state: "acquired" as const, cachedResponse: undefined },
        });
      }

      // 2. Key exists — check if it's a sentinel or a cached response.
      const getResult = await this.get.handleAsync({ key: cacheKey });

      if (
        !getResult.success ||
        getResult.data?.value === undefined ||
        getResult.data.value === null
      ) {
        // GET failed or returned null — key may have expired between SET NX and GET.
        // Fail-open: treat as acquired.
        this.context.logger.warn(
          `Idempotency key exists but GET returned no value. Failing open. TraceId: ${this.traceId}`,
        );
        return D2Result.ok({
          data: { state: "acquired" as const, cachedResponse: undefined },
        });
      }

      // 3. Check if sentinel (in-flight) or cached response.
      if (getResult.data.value === SENTINEL) {
        return D2Result.ok({
          data: { state: "in_flight" as const, cachedResponse: undefined },
        });
      }

      // 4. Try to parse as CachedResponse.
      try {
        const cachedResponse = JSON.parse(getResult.data.value) as Idempotency.CachedResponse;
        if (cachedResponse && typeof cachedResponse.statusCode === "number") {
          return D2Result.ok({
            data: { state: "cached" as const, cachedResponse },
          });
        }
      } catch {
        this.context.logger.warn(
          `Failed to deserialize cached response for idempotency key. Failing open. TraceId: ${this.traceId}`,
        );
      }

      // Could not parse — fail-open: treat as acquired.
      return D2Result.ok({
        data: { state: "acquired" as const, cachedResponse: undefined },
      });
    } catch {
      // Fail-open on all cache errors.
      this.context.logger.warn(`Idempotency check failed. Failing open. TraceId: ${this.traceId}`);
      return D2Result.ok({
        data: { state: "acquired" as const, cachedResponse: undefined },
      });
    }
  }
}

File: backends/dotnet/shared/Implementations/Middleware/Idempotency.Default/Handlers/X/Check.cs
// -----------------------------------------------------------------------
// <copyright file="Check.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Idempotency.Default.Handlers.X;

using System.Text.Json;
using D2.Shared.Handler;
using D2.Shared.Interfaces.Caching.Distributed.Handlers.C;
using D2.Shared.Interfaces.Caching.Distributed.Handlers.R;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using H = D2.Shared.Idempotency.Default.Interfaces.IIdempotency.ICheckHandler;
using I = D2.Shared.Idempotency.Default.Interfaces.IIdempotency.CheckInput;
using O = D2.Shared.Idempotency.Default.Interfaces.IIdempotency.CheckOutput;

/// <summary>
/// Handler for checking idempotency state using SET NX + GET pattern.
/// </summary>
/// <remarks>
/// Complex handler (X/) because it may mutate shared state (creates Redis keys)
/// while its primary intent is retrieval of state.
/// </remarks>
public class Check : BaseHandler<Check, I, O>, H
{
    private const string _SENTINEL = "__processing__";
    private const string _KEY_PREFIX = "idempotency:";

    private readonly ICreate.ISetNxHandler<string> r_setNx;
    private readonly IRead.IGetHandler<string> r_get;
    private readonly IdempotencyOptions r_options;

    /// <summary>
    /// Initializes a new instance of the <see cref="Check"/> class.
    /// </summary>
    ///
    /// <param name="setNx">
    /// The handler for SET NX operations.
    /// </param>
    /// <param name="get">
    /// The handler for GET operations.
    /// </param>
    /// <param name="options">
    /// The idempotency options.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Check(
        ICreate.ISetNxHandler<string> setNx,
        IRead.IGetHandler<string> get,
        IOptions<IdempotencyOptions> options,
        IHandlerContext context)
        : base(context)
    {
        r_setNx = setNx;
        r_get = get;
        r_options = options.Value;
    }

    /// <inheritdoc />
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        var cacheKey = $"{_KEY_PREFIX}{input.IdempotencyKey}";

        try
        {
            // 1. Attempt to acquire the lock with SET NX.
            var setNxResult = await r_setNx.HandleAsync(
                new ICreate.SetNxInput<string>(cacheKey, _SENTINEL, r_options.InFlightTtl),
                ct);

            if (setNxResult.CheckSuccess(out var setNxOutput) && setNxOutput?.WasSet == true)
            {
                // Lock acquired — caller should proceed with request processing.
                return D2Result<O?>.Ok(
                    new O(IdempotencyState.Acquired, null));
            }

            // 2. Key exists — check if it's a sentinel or a cached response.
            var getResult = await r_get.HandleAsync(
                new IRead.GetInput(cacheKey),
                ct);

            if (!getResult.CheckSuccess(out var getOutput) || getOutput?.Value is null)
            {
                // GET failed or returned null — key may have expired between SET NX and GET.
                // Fail-open: treat as acquired.
                Context.Logger.LogWarning(
                    "Idempotency key exists but GET returned no value. Failing open. TraceId: {TraceId}",
                    TraceId);

                return D2Result<O?>.Ok(
                    new O(IdempotencyState.Acquired, null));
            }

            // 3. Check if sentinel (in-flight) or cached response.
            if (getOutput.Value == _SENTINEL)
            {
                return D2Result<O?>.Ok(
                    new O(IdempotencyState.InFlight, null));
            }

            // 4. Try to parse as CachedResponse.
            try
            {
                var cachedResponse = JsonSerializer.Deserialize<CachedResponse>(getOutput.Value);
                if (cachedResponse is not null)
                {
                    return D2Result<O?>.Ok(
                        new O(IdempotencyState.Cached, cachedResponse));
                }
            }
            catch (JsonException ex)
            {
                Context.Logger.LogWarning(
                    ex,
                    "Failed to deserialize cached response for idempotency key. Failing open. TraceId: {TraceId}",
                    TraceId);
            }

            // Could not parse — fail-open: treat as acquired.
            return D2Result<O?>.Ok(
                new O(IdempotencyState.Acquired, null));
        }
        catch (Exception ex)
        {
            // Fail-open on all cache errors.
            Context.Logger.LogWarning(
                ex,
                "Idempotency check failed. Failing open. TraceId: {TraceId}",
                TraceId);

            return D2Result<O?>.Ok(
                new O(IdempotencyState.Acquired, null));
        }
    }
}
